toc/rfc/rfc-draft-ssh-over-ws.md

```diff
@@ -0,0 +1,270 @@
+
+#### Scenario C: WS only
+This scenario MAY be enabled by operators. It turns of the legacy scenario A and allows operators to close the public port 2222 on the load balancer. The CF CLI MUST understand the new flow using a WebSocket tunnel. Other components MUST support the new flow and MAY support both scenarios.
+
+
+### Changes per Component (Scenario B and C)
+
+#### CF CLI
+- The CLI MUST query the "v2/info" endpoint of Cloud Controller.
+- The endpoint MUST return the legacy SSH info:
+
+```
+  "app_ssh_endpoint": "ssh.cf-app.com:2222",
+  "app_ssh_host_key_fingerprint": "TD1dRQINLi2KxilVLzAI8tXB2h8MP79oyVJnUwshjdc",
+  "app_ssh_oauth_client": "ssh-proxy",
+```
+
+- The endpoint MAY return an additional field for ssh-over-websocket:
+
+```
+  "app_ssh_ws_endpoint": "wss://ssh.cf-app.com",
+```
+
+- CF CLI MUST prefer the `app_ssh_ws_endpoint` over the `app_ssh_endpoint` if present.
+- CF CLI MAY fall back to `app_ssh_endpoint` if `app_ssh_ws_endpoint` did not connect successfully.
+- When using `app_ssh_ws_endpoint`, the CF CLI MUST wrap the SSH connection in a WebSocket connection:
+
+```
+func (c *secureShell) Connect(opts *options.SSHOptions) error {
+	err := c.validateTarget(opts)
+	if err != nil {
+		return err
+	}
+
+	clientConfig := &ssh.ClientConfig{
+		User: fmt.Sprintf("cf:%s/%d", c.app.GUID, opts.Index),
+		Auth: []ssh.AuthMethod{
+			ssh.Password(c.token),
+		},
+		HostKeyCallback: fingerprintCallback(opts, c.sshEndpointFingerprint),
+	}
+
+	// Wrap SSH in WebSocket if possible
+	var secureClient SecureClient
+	if c.sshWsEndpoint != "" {
+		secureClient, err = tunnelSSHThruWebSocket(c.sshWsEndpoint, clientConfig)
+	}
+
+	if err != nil {
+		// Fall back to SSH over TCP
+		secureClient, err = c.secureDialer.Dial("tcp", c.sshEndpoint, clientConfig)
+		if err != nil {
+			return err
+		}
+	}
+
+	c.secureClient = secureClient
+	c.opts = opts
+	return nil
+}
+```
+Code adapted from [cloudfoundry/cli ssh.go](https://github.com/cloudfoundry/cli/blob/main/cf/ssh/ssh.go#L119)
+
+The `sshWsEndpoint` would have to be fetched from the `app_ssh_ws_endpoint` field of the "info" endpoint of CAPI.
+
+
+A `tunnelSSHThruWebSocket` function could look something like this:
+```
+func tunnelSSHThruWebSocket(url string, config *ssh.ClientConfig) (SecureClient, error) {
+	wsConn, err := websocket.Dial(url, "", url)
+	if err != nil {
+		return nil, err
+	}
+
+	c, chans, reqs, err := ssh.NewClientConn(wsConn, url, config)
+	if err != nil {
+		return nil, err
+	}
+	sshClient := ssh.NewClient(c, chans, reqs)
+	return &secureClient{client: sshClient}, nil
+}
+```
+
+- The rest of CF CLI SHOULD be unchanged and work as before.
+
+#### CAPI
+
+- CAPI MAY provide an additional `app_ssh_ws_endpoint` field in the "v2/info" endpoint:
+```
+  "app_ssh_ws_endpoint": "wss://ssh.cf-app.com"
+```
+
+- The URL presented in the field MUST be `wss://` + the route as announced by route-registrar for SSH proxy (see [CF Deployment](#cf-deployment))
```

Comment:
would it be useful to have an addressing scheme that is directly related to the app's route?

e.g. `cf ssh wss://myapp.cf-app.com`? The fact that this is an SSH tunnel request could be indicated via separate header. Ultimately the websocket communication will end up in gorouter, so gorouter can decide that this is actually a web service request. Finally, we also need to communicate the instance ID that we want to connect to, which would also be in a header.

Reply:
The web socket tunnel breaks with some of the well-known concepts. Previously there was just one SSH server, the ssh-scheduler (or whatever it's called) and it decides to which app you will be connected based on the user (see also the [docs](https://docs.cloudfoundry.org/devguide/deploy-apps/ssh-apps.html#other-ssh-access)). Now we have this web socket tunnel and in theory you still have to say where you want to go once you exit the tunnel, but you don't want to go anywhere. So the real target host for ssh is probably going to be `localhost`?

`ssh -o ProxyCommand=cf-ws-tunnel cf:${process_id}/0@$host`

* `cf-ws-tunnel` would talk to the CC and get `wss://ssh.cf.example.com` as the WebSocket endpoint which points to the SSH server (registered via route-registrar like any other route is).
* `cf:${process_id}/0` is the same as it is right now, tells the SSH server which process to access.
* `$host`: it's unclear to me what that value would be, I'm suspecting it has to be `localhost` or it doesn't matter at all since the web socket tunnel will be piped to the SSH server no matter what.

It becomes this weird hybrid of a tunnel and an SSH connection which both are taking on responsibilities of the other.

Reply:
I disagree. The point of this feature request is to access SSH proxy via a WebSocket tunnel instead of a direct SSH connection. Changing the way CF SSH works in general is out of scope of this RFC. For gorouter, this will be just another WebSocket request without any special handling needed.